sumcheck/src/transcript.rs

```diff
@@ -0,0 +1,113 @@
+use crate::{
+    fieldbinding::{FromFieldBinding, ToFieldBinding},
+    GPUApiWrapper, LibraryError,
+};
+use cudarc::driver::CudaSlice;
+use ff::PrimeField;
+use halo2curves::serde::SerdeObject;
+use std::io::{Cursor, Read, Write};
+
+pub enum Hash {
+    Keccack256,
+}
+
+pub trait CudaTranscript<F: PrimeField + ToFieldBinding<F> + FromFieldBinding<F>> {
+    fn get_hash_method(&self) -> Hash;
+    fn get_cuda_slice(
+        &mut self,
+        gpu: &mut GPUApiWrapper<F>,
+        count: usize,
+        add_len: usize,
+    ) -> Result<TranscriptInner, LibraryError>;
+}
+
+pub struct TranscriptInner {
+    pub start: CudaSlice<u8>,
+    pub cursor: usize,
+    pub end: usize,
+}
+
+impl TranscriptInner {
+    fn new(start: CudaSlice<u8>, cursor: usize, end: usize) -> Self {
+        Self { start, cursor, end }
+    }
+}
+
+pub fn from_u8_to_f<F: PrimeField + SerdeObject>(v: &[u8]) -> F {
+    F::from_raw_bytes_unchecked(v) * F::ONE
+}
+
+pub struct CudaKeccakTranscript<F> {
```

Comment:
So I think we should declare `CudaKeccakTranscript` in C++ side, and let `bindgen` to generate Rust struct automatically for us. `FieldBinding` works in this way, would you check this?